File: ros_gz_sim/src/gzserver.cpp
// Copyright 2024 Open Source Robotics Foundation, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "ros_gz_sim/gzserver.hpp"

#include <functional>
#include <memory>
#include <string>
#include <thread>

#include <gz/common/Console.hh>
#include <gz/sim/Server.hh>
#include <gz/sim/ServerConfig.hh>
#include <gz/sim/SystemLoader.hh>
#include <rclcpp/rclcpp.hpp>
#include <rclcpp/utilities.hpp>
#include <rclcpp_components/register_node_macro.hpp>

#include "ros_gz_sim/gz_simulation_interfaces.hpp"

namespace ros_gz_sim
{

class GzServer::Implementation
{
  /// \brief We don't want to block the ROS thread.

public:
  std::thread thread;
  std::unique_ptr<gz_simulation_interfaces::GzSimulationInterfaces> sim_interfaces;
};

GzServer::GzServer(const rclcpp::NodeOptions & options)
: Node("gzserver", options), dataPtr(gz::utils::MakeUniqueImpl<Implementation>())
{
  rclcpp::uninstall_signal_handlers();
  this->dataPtr->thread = std::thread(std::bind(&GzServer::OnStart, this));
}

GzServer::~GzServer()
{
  // Make sure to join the thread on shutdown.
  if (this->dataPtr->thread.joinable()) {
    this->dataPtr->thread.join();
  }
}

void GzServer::OnStart()
{
  auto world_sdf_file = this->declare_parameter("world_sdf_file", "");
  auto world_sdf_string = this->declare_parameter("world_sdf_string", "");
  auto initial_sim_time = this->declare_parameter("initial_sim_time", 0.0);
  const auto verbosity_level = this->declare_parameter("verbosity_level", 4);

  gz::common::Console::SetVerbosity(verbosity_level);
  gz::sim::ServerConfig server_config;

  if (!world_sdf_file.empty()) {
    server_config.SetSdfFile(world_sdf_file);
  } else if (!world_sdf_string.empty()) {
    server_config.SetSdfString(world_sdf_string);
  } else {
    RCLCPP_ERROR(this->get_logger(), "Must specify either 'world_sdf_file' or 'world_sdf_string'");
    rclcpp::shutdown();
    return;
  }
  server_config.SetInitialSimTime(initial_sim_time);

  auto server = std::make_unique<gz::sim::Server>(server_config);
  // TODO(azeey) Think about whether it makes sense to RunOnce paused here or wait for some
  // critical services from Gazebo to become available before starting the ROS services
  server->RunOnce(true);
  // TODO(azeey) Allow disabling simulation interfaces

  this->dataPtr->sim_interfaces =
    std::make_unique<gz_simulation_interfaces::GzSimulationInterfaces>(
      this->create_sub_node(this->get_name()));
  server->Run(true /*blocking*/, 0, false /*paused*/);
  server.reset();
  // Call shutdown before resetting sim_interfaces so that threads in sim_interfaces can gracefully
  // exit before being destructed.
  rclcpp::shutdown();
  this->dataPtr->sim_interfaces.reset();
}

}  // namespace ros_gz_sim

RCLCPP_COMPONENTS_REGISTER_NODE(ros_gz_sim::GzServer)

File: launch_gz/launch_gz/actions/gzserver.py
# Copyright 2024 Open Source Robotics Foundation, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Module for the GzServer action."""

from typing import List
from typing import Optional

from launch.action import Action
from launch.actions import IncludeLaunchDescription
from launch.frontend import expose_action, Entity, Parser
from launch.launch_context import LaunchContext
from launch.launch_description_sources import PythonLaunchDescriptionSource
from launch.some_substitutions_type import SomeSubstitutionsType
from launch.substitutions import PathJoinSubstitution
from launch_ros.substitutions import FindPackageShare

@expose_action('gzserver')
class GzServer(Action):
	"""Action that executes a gzserver ROS [composable] node."""

	def __init__(
		self,
		*,
		world_sdf_file: SomeSubstitutionsType,
		world_sdf_string: SomeSubstitutionsType,
		container_name: SomeSubstitutionsType,
		use_composition: SomeSubstitutionsType,
		**kwargs
	) -> None:
		"""
		Construct a gzserver action.

		All arguments are forwarded to `ros_gz_sim.launch.gzserver.launch.py`, so see the documentation
		of that class for further details.

		:param: world_sdf_file Path to the SDF world file.
		:param: world_sdf_string SDF world string.
		:param: container_name Name of container that nodes will load in if use composition.
		:param: use_composition Use composed bringup if True.
		"""

		super().__init__(**kwargs)
		self.__world_sdf_file = world_sdf_file
		self.__world_sdf_string = world_sdf_string
		self.__container_name = container_name
		self.__use_composition = use_composition

	@classmethod
	def parse(cls, entity: Entity, parser: Parser):
		"""Parse gzserver."""
		_, kwargs = super().parse(entity, parser)

		world_sdf_file = entity.get_attr(
			'world_sdf_file', data_type=str,
			optional=True)

		world_sdf_string = entity.get_attr(
		    'world_sdf_string', data_type=str,
		    optional=True)

		container_name = entity.get_attr(
		    'container_name', data_type=str,
		    optional=True)

		use_composition = entity.get_attr(
		    'use_composition', data_type=str,
		    optional=True)

		if isinstance(world_sdf_file, str):
			world_sdf_file = parser.parse_substitution(world_sdf_file)
			kwargs['world_sdf_file'] = world_sdf_file

		if isinstance(world_sdf_string, str):
		    world_sdf_string = parser.parse_substitution(world_sdf_string)
		    kwargs['world_sdf_string'] = world_sdf_string

		if isinstance(container_name, str):
		    container_name = parser.parse_substitution(container_name)
		    kwargs['container_name'] = container_name

		if isinstance(use_composition, str):
		    use_composition = parser.parse_substitution(use_composition)
		    kwargs['use_composition'] = use_composition

		return cls, kwargs

	def execute(self, context: LaunchContext) -> Optional[List[Action]]:
		"""
		Execute the action.
		"""

		gzserver_description = IncludeLaunchDescription(
			PythonLaunchDescriptionSource(
				[PathJoinSubstitution([FindPackageShare('ros_gz_sim'),
									   'launch',
									   'gzserver.launch.py'])]),
				launch_arguments=[('world_sdf_file', self.__world_sdf_file),
								  ('world_sdf_string', self.__world_sdf_string),
								  ('container_name',   self.__container_name),
								  ('use_composition',  self.__use_composition),
								 ])

		return [gzserver_description]
